compiler/rustc_expand/src/proc_macro_server.rs

```diff
@@ -421,13 +427,58 @@ impl server::TokenStream for Rustc<'_> {
         parse_stream_from_source_str(
             FileName::proc_macro_source_code(src),
             src.to_string(),
-            self.sess,
+            self.sess(),
             Some(self.call_site),
         )
     }
     fn to_string(&mut self, stream: &Self::TokenStream) -> String {
         pprust::tts_to_string(stream)
     }
+    fn expand_literal(&mut self, stream: &Self::TokenStream) -> Result<Self::Literal, ()> {
+        // Parse the expression from our tokenstream.
+        let expr = {
+            let mut p = rustc_parse::stream_to_parser(
+                self.sess(),
+                stream.clone(),
+                Some("proc_macro expand literal"),
+            );
+            let expr = p.parse_expr().map_err(|_| ())?;
+            if p.token != token::Eof {
+                return Err(());
+            }
+            expr
+        };
+
+        // Perform eager expansion on the expression.
+        let expr = self
+            .ecx
+            .expander()
+            .fully_expand_fragment(crate::expand::AstFragment::Expr(expr))
+            .make_expr();
+
+        // Extract the literal value from the expression.
+        let lit = match &expr.kind {
+            ast::ExprKind::Lit(l) => l.kind.to_lit_token(),
+            ast::ExprKind::Unary(ast::UnOp::Neg, e) => match &e.kind {
```

Comment:
I would prefer to avoid hacks like this, `-1` is not a single token and `token::Lit` is supposed to never keep anything with `-`.

Reply:
My reasoning behind having this hack was that a `-1` `token::Lit` already exists whenever the stable `Literal::i32_unsuffixed(-1)` API is called (https://doc.rust-lang.org/nightly/proc_macro/struct.Literal.html#method.i32_unsuffixed). This change is consistent with that behaviour, even if it's quite unfortunate. I don't think there's any way we can get away at this point with not supporting negative `Literal` objects in the proc_macro API.

These weird tokens are broken apart before being exposed to the rest of rustc (https://searchfox.org/rust/rev/7c89e389d00cfc7e86ae7e1b45880da4f5f5c9f5/compiler/rustc_expand/src/proc_macro_server.rs#217,228), and with the changes in #86822 won't even be a `token::Lit` anymore.

Reply:
Ok, `i32_unsuffixed` and friends are indeed packing the minus into `token::Lit`. That's unfortunate.

Reply:
If we wanted, we could change the return type of `expand_literal` to be `TokenStream`. This would let us represent a negative numeric literal properly as two tokens, and would also allow us to potentially support boolean `true` and `false` keyword literals, as they need to be stored as an `Ident` rather than `Literal`.

I mostly used `Literal` as the return type because it seemed like the obvious choice, and it might be confusing that `TokenStream` was returned after expanding to a literal :-)

Reply:
If `i32_unsuffixed` can produce a `Literal` with minus in `Symbol`, then this is acceptable too.
Although it it would be better to use some representation not requiring keeping the minus in the string (like keeping a private boolean flag instead), perhaps it can be done as a part of #86822.

Reply:
Sure, that wouldn't be a hard thing to implement.